File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    static ArrayList<Integer> graph[];
    static boolean visited[];
    public int solution(int n, int[][] wires) {
        int answer = Integer.MAX_VALUE;
        graph = new ArrayList[n+1];
        for(int i = 0; i <= n; i++){
            graph[i] = new ArrayList<>();
        }
        for(int[] wire : wires){
            int from = wire[0];
            int to = wire[1];
            graph[from].add(to);
            graph[to].add(from);
        }
        // for(int k = 1; k < n; k++)
            // System.out.println(k+ " " + graph[k].toString());
        int count1 = 0;
        int count2 = 0;
        for(int[] wire: wires){
            int from = wire[0];
            int to = wire[1];
            count1 = bfs(from, to, n);
            count2 = bfs(to, from, n);
            // System.out.println(count1);
            // System.out.println(count2);
            answer = Math.min(answer, Math.abs(count1-count2));
        }
        return answer;
    }
    public int bfs(int from, int to, int size){
        int count = 0;
        visited = new boolean[size+1];
        Queue<Integer> queue = new ArrayDeque<>();
        queue.offer(from);
        while(!queue.isEmpty()){
            int cur = queue.poll();
            for(int next : graph[cur]){
                if (visited[next]) continue;
                visited[next] = true;
                if (!check(next,cur, from, to)) continue;
                queue.offer(next);
                count++;
            }
        }
        return count==0?1:count;
        // return count;
    }
    public boolean check(int next, int cur, int from, int to){
        // System.out.println("next = " + next + " cur = " + cur + " from = " + from + " to = " + to);
        if( (next == from && cur == to) || (next == to && cur == from)) return false;
        return true;
    } 
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque
N = 0
def solution(n, wires):
    global N
    N = n
    visited =[False for i in range(n + 1)]
    answer = 99999
    graph = [ [] for i in range(n + 1)]
    for s, e in wires:
        graph[s].append(e)
        graph[e].append(s)
    for s, e in wires:
        graph[s].remove(e)
        graph[e].remove(s)
        # print(graph)
        a = bfs(s, graph)
        b = bfs(e, graph)
        # print("a ",a,"b",b)
        answer = min(answer, abs(a - b))
        graph[s].append(e)
        graph[e].append(s)
        pass
    return answer

def bfs(start, graph):
    # print("start = ",start)
    visited = [0 for i in range(N + 1)]
    queue = deque()
    queue.append(start)
    while queue:
        # print("queue = ", queue)
        cur = queue.popleft()
        # print("cur = ",cur)
        if visited[cur] == 1:
            continue
        visited[cur] = 1
        for node in graph[cur]:
            queue.append(node)
            # visited[node] = 1
    # print(visited)
    return sum(visited)
    